File: api/akademik.php
<?php
// ============================================
// Akademik API Handlers (KRS, KHS, Jadwal, Ujian)
// Port of Go akademik.go → PHP + MySQL
// ============================================

/**
 * GET /api/akademik/krs/:nim
 */
function getKRS($nim) {
    $db = getDB();
    $semester = $_GET['semester'] ?? '';

    $sql = "SELECT k.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks, mk.semester as mk_semester
            FROM krs k LEFT JOIN mata_kuliah mk ON k.mata_kuliah_id = mk.id
            WHERE k.nim = ?";
    $params = [$nim];

    if ($semester) {
        $sql .= " AND k.semester = ?";
        $params[] = $semester;
    }
    $sql .= " ORDER BY k.id ASC";

    $stmt = $db->prepare($sql);
    $stmt->execute($params);
    $items = $stmt->fetchAll();

    $totalSks = 0;
    $formatted = array_map(function($row) use (&$totalSks) {
        $sks = (int)($row['mk_sks'] ?? 0);
        $totalSks += $sks;
        return [
            'id' => (int)$row['id'], 'nim' => $row['nim'], 'semester' => (int)$row['semester'],
            'tahun_ajaran' => $row['tahun_ajaran'], 'mata_kuliah_id' => (int)$row['mata_kuliah_id'],
            'kelas' => $row['kelas'], 'seksi' => $row['seksi'], 'dosen_nama' => $row['dosen_nama'],
            'waktu' => $row['waktu'], 'ruang' => $row['ruang'], 'jenis_kelas' => $row['jenis_kelas'],
            'status' => $row['status'], 'disetujui_oleh' => $row['disetujui_oleh'],
            'mata_kuliah' => ['id' => (int)$row['mata_kuliah_id'], 'kode' => $row['mk_kode'], 'nama' => $row['mk_nama'], 'sks' => $sks, 'semester' => (int)($row['mk_semester'] ?? 0)],
        ];
    }, $items);

    jsonResponse(['nim' => $nim, 'semester' => $semester, 'total_sks' => $totalSks, 'count' => count($formatted), 'data' => $formatted]);
}

/**
 * GET /api/akademik/khs/:nim
 */
function getKHS($nim) {
    $db = getDB();
    $semester = $_GET['semester'] ?? '';

    $sql = "SELECT n.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
            FROM nilai n LEFT JOIN mata_kuliah mk ON n.mata_kuliah_id = mk.id
            WHERE n.nim = ?";
    $params = [$nim];

    if ($semester) {
        $sql .= " AND n.semester = ?";
        $params[] = $semester;
    }
    $sql .= " ORDER BY n.semester ASC, n.id ASC";

    $stmt = $db->prepare($sql);
    $stmt->execute($params);
    $items = $stmt->fetchAll();

    $totalSks = 0;
    $totalBobot = 0.0;
    $grouped = [];

    $formatted = array_map(function($row) use (&$totalSks, &$totalBobot, &$grouped) {
        $sks = (int)($row['mk_sks'] ?? 0);
        $bobot = (float)$row['bobot'];
        $totalSks += $sks;
        $totalBobot += $sks * $bobot;
        $sem = (int)$row['semester'];

        $item = [
            'id' => (int)$row['id'], 'nim' => $row['nim'], 'mata_kuliah_id' => (int)$row['mata_kuliah_id'],
            'semester' => $sem, 'nilai' => $row['nilai'], 'bobot' => $bobot, 'kelas' => $row['kelas'],
            'mata_kuliah' => ['id' => (int)$row['mata_kuliah_id'], 'kode' => $row['mk_kode'], 'nama' => $row['mk_nama'], 'sks' => $sks],
        ];
        $grouped[$sem][] = $item;
        return $item;
    }, $items);

    $ipk = $totalSks > 0 ? round($totalBobot / $totalSks, 2) : 0;

    jsonResponse([
        'nim' => $nim, 'total_sks' => $totalSks, 'total_bobot' => $totalBobot,
        'ipk' => $ipk, 'count' => count($formatted), 'data' => $formatted, 'per_semester' => $grouped,
    ]);
}

/**
 * GET /api/akademik/jadwal/:nim
 */
function getJadwal($nim) {
    $db = getDB();
    $semester = $_GET['semester'] ?? '';

    $sql = "SELECT k.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
            FROM krs k LEFT JOIN mata_kuliah mk ON k.mata_kuliah_id = mk.id
            WHERE k.nim = ?";
    $params = [$nim];

    if ($semester) {
        $sql .= " AND k.semester = ?";
        $params[] = $semester;
    }
    $sql .= " ORDER BY k.waktu ASC";

    $stmt = $db->prepare($sql);
    $stmt->execute($params);
    $items = $stmt->fetchAll();

    $formatted = array_map(function($row) {
        return [
            'id' => (int)$row['id'], 'nim' => $row['nim'], 'semester' => (int)$row['semester'],
            'mata_kuliah_id' => (int)$row['mata_kuliah_id'], 'kelas' => $row['kelas'],
            'dosen_nama' => $row['dosen_nama'], 'waktu' => $row['waktu'], 'ruang' => $row['ruang'],
            'mata_kuliah' => ['id' => (int)$row['mata_kuliah_id'], 'kode' => $row['mk_kode'], 'nama' => $row['mk_nama'], 'sks' => (int)($row['mk_sks'] ?? 0)],
        ];
    }, $items);

    jsonResponse(['nim' => $nim, 'count' => count($formatted), 'data' => $formatted]);
}

/**
 * GET /api/akademik/ujian/:nim
 */
function getJadwalUjian($nim) {
    $db = getDB();

    $stmt = $db->prepare('SELECT mata_kuliah_id FROM krs WHERE nim = ? ORDER BY semester DESC');
    $stmt->execute([$nim]);
    $mkIds = $stmt->fetchAll(PDO::FETCH_COLUMN);

    if (empty($mkIds)) {
        jsonResponse(['nim' => $nim, 'count' => 0, 'data' => []]);
    }

    $placeholders = implode(',', array_fill(0, count($mkIds), '?'));
    $stmt = $db->prepare("
        SELECT ju.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
        FROM jadwal_ujian ju LEFT JOIN mata_kuliah mk ON ju.mata_kuliah_id = mk.id
        WHERE ju.mata_kuliah_id IN ($placeholders) ORDER BY ju.tanggal ASC
    ");
    $stmt->execute($mkIds);
    $items = $stmt->fetchAll();

    $formatted = array_map(function($row) {
        return [
            'id' => (int)$row['id'], 'mata_kuliah_id' => (int)$row['mata_kuliah_id'],
            'tanggal' => $row['tanggal'], 'mulai' => $row['mulai'], 'selesai' => $row['selesai'],
            'sesi' => $row['sesi'], 'jenis' => $row['jenis'], 'kelompok' => (int)$row['kelompok'],
            'mata_kuliah' => ['id' => (int)$row['mata_kuliah_id'], 'kode' => $row['mk_kode'], 'nama' => $row['mk_nama'], 'sks' => (int)($row['mk_sks'] ?? 0)],
        ];
    }, $items);

    jsonResponse(['nim' => $nim, 'count' => count($formatted), 'data' => $formatted]);
}

/**
 * GET /api/akademik/matakuliah
 */
function getAllMataKuliah() {
    $db = getDB();
    $stmt = $db->query('SELECT * FROM mata_kuliah ORDER BY kode ASC');
    $items = $stmt->fetchAll();

    $formatted = array_map(function($row) {
        return [
            'id' => (int)$row['id'], 'kode' => $row['kode'], 'nama' => $row['nama'],
            'sks' => (int)$row['sks'], 'semester' => (int)$row['semester'], 'prodi' => $row['prodi'],
        ];
    }, $items);

    jsonResponse(['count' => count($formatted), 'data' => $formatted]);
}

File: backend/handlers/akademik.go
package handlers

import (
	"net/http"

	"stia-bayuangga/backend/config"
	"stia-bayuangga/backend/models"

	"github.com/gin-gonic/gin"
)

// ===================== KRS =====================

// GetKRS returns the student's KRS for a given semester
// GET /api/akademik/krs/:nim?semester=4
func GetKRS(c *gin.Context) {
	nim := c.Param("nim")
	semester := c.Query("semester")

	var items []models.KRS
	q := config.DB.Preload("MataKuliah").Where("nim = ?", nim)
	if semester != "" {
		q = q.Where("semester = ?", semester)
	}
	q.Order("id asc").Find(&items)

	// Compute total SKS
	totalSks := 0
	for _, item := range items {
		totalSks += item.MataKuliah.SKS
	}

	c.JSON(http.StatusOK, gin.H{
		"nim":       nim,
		"semester":  semester,
		"total_sks": totalSks,
		"count":     len(items),
		"data":      items,
	})
}

// ===================== KHS =====================

// GetKHS returns graded courses per semester
// GET /api/akademik/khs/:nim?semester=3
func GetKHS(c *gin.Context) {
	nim := c.Param("nim")
	semester := c.Query("semester")

	var nilaiList []models.Nilai
	q := config.DB.Preload("MataKuliah").Where("nim = ?", nim)
	if semester != "" {
		q = q.Where("semester = ?", semester)
	}
	q.Order("semester asc, id asc").Find(&nilaiList)

	totalSks := 0
	totalBobot := 0.0
	for _, n := range nilaiList {
		totalSks += n.MataKuliah.SKS
		totalBobot += float64(n.MataKuliah.SKS) * n.Bobot
	}
	ipk := 0.0
	if totalSks > 0 {
		ipk = totalBobot / float64(totalSks)
	}

	// Group by semester
	grouped := map[int][]models.Nilai{}
	for _, n := range nilaiList {
		grouped[n.Semester] = append(grouped[n.Semester], n)
	}

	c.JSON(http.StatusOK, gin.H{
		"nim":         nim,
		"total_sks":   totalSks,
		"total_bobot": totalBobot,
		"ipk":         ipk,
		"count":       len(nilaiList),
		"data":        nilaiList,
		"per_semester": grouped,
	})
}

// ===================== JADWAL =====================

// GetJadwal returns the student's weekly schedule from KRS
// GET /api/akademik/jadwal/:nim
func GetJadwal(c *gin.Context) {
	nim := c.Param("nim")
	semester := c.Query("semester")

	var items []models.KRS
	q := config.DB.Preload("MataKuliah").Where("nim = ?", nim)
	if semester != "" {
		q = q.Where("semester = ?", semester)
	}
	q.Order("waktu asc").Find(&items)

	c.JSON(http.StatusOK, gin.H{
		"nim":   nim,
		"count": len(items),
		"data":  items,
	})
}

// GetJadwalUjian returns UTS/UAS exam schedules
// GET /api/akademik/ujian/:nim
func GetJadwalUjian(c *gin.Context) {
	nim := c.Param("nim")

	// Get MataKuliah IDs from KRS for latest semester
	var krsItems []models.KRS
	config.DB.Where("nim = ?", nim).Order("semester desc").Find(&krsItems)

	if len(krsItems) == 0 {
		c.JSON(http.StatusOK, gin.H{"nim": nim, "count": 0, "data": []models.JadwalUjian{}})
		return
	}

	// Get MK IDs
	mkIDs := []uint{}
	for _, k := range krsItems {
		mkIDs = append(mkIDs, k.MataKuliahID)
	}

	var ujianList []models.JadwalUjian
	config.DB.Preload("MataKuliah").Where("mata_kuliah_id IN ?", mkIDs).Order("tanggal asc").Find(&ujianList)

	c.JSON(http.StatusOK, gin.H{
		"nim":   nim,
		"count": len(ujianList),
		"data":  ujianList,
	})
}

// ===================== MATA KULIAH =====================

// GetAllMataKuliah returns the full course catalog
// GET /api/akademik/matakuliah
func GetAllMataKuliah(c *gin.Context) {
	var mkList []models.MataKuliah
	config.DB.Order("kode asc").Find(&mkList)
	c.JSON(http.StatusOK, gin.H{"count": len(mkList), "data": mkList})
}
